engine/src/conversion/analysis/fun/implicit_constructors.rs

```diff
@@ -12,308 +12,645 @@
+                    non_const_copy_constructor: is_explicit(ExplicitKind::NonConstCopyConstructor),
+                    move_constructor: is_explicit(ExplicitKind::MoveConstructor),
+                };
+                log::info!(
+                    "Special member functions found for {:?}: {:?}",
+                    name,
+                    items_found
+                );
+                assert!(
+                    all_items_found.insert(name.clone(), items_found).is_none(),
+                    "Duplicate struct: {:?}",
+                    name
+                );
+            } else {
+                // If no user-declared constructors of any kind are provided for a class type (struct, class, or union),
+                // the compiler will always declare a default constructor as an inline public member of its class.
+                //
+                // The implicitly-declared or defaulted default constructor for class T is defined as deleted if any of the following is true:
+                // T has a member of reference type without a default initializer.
+                // T has a non-const-default-constructible const member without a default member initializer.
+                // T has a member (without a default member initializer) which has a deleted default constructor, or its default constructor is ambiguous or inaccessible from this constructor.
+                // T has a direct or virtual base which has a deleted default constructor, or it is ambiguous or inaccessible from this constructor.
+                // T has a direct or virtual base or a non-static data member which has a deleted destructor, or a destructor that is inaccessible from this constructor.
+                // T is a union with at least one variant member with non-trivial default constructor, and no variant member of T has a default member initializer. // we don't support unions anyway
+                // T is a non-union class with a variant member M with a non-trivial default constructor, and no variant member of the anonymous union containing M has a default member initializer.
+                // T is a union and all of its variant members are const. // we don't support unions anyway
+                //
+                // Variant members are the members of anonymous unions.
+                let default_constructor = {
+                    let explicit = find_explicit(ExplicitKind::DefaultConstructor);
+                    // TODO: For https://github.com/google/autocxx/issues/815, replace the first term with:
+                    //   explicit.map_or(true, |explicit_found| matches!(explicit_found, ExplicitFound::Defaulted(_)))
+                    let have_defaulted = explicit.is_none()
+                        && !explicits.iter().any(|(ExplicitType { ty, kind }, _)| {
+                            ty == name
+                                && match *kind {
+                                    ExplicitKind::DefaultConstructor => false,
```

Comment:
Nit: I think this would be clearer as `matches!(*kind, ExplicitKind::ConstCopyConstructor | ExplicitKind::NonConstCopyConstructor | etc.)`

Reply:
I want to force new `ExplicitKind`s to be added here. For example, splitting out more specific varieties of move or copy constructor for special handling would need different decisions, but using `matches!` would implicitly treat them as not removing the default constructor.

Reply:
Ah. OK. Good call.